### DISCUSSION TITLE
About local ccache

### DISCUSSION THREAD

**Discussion by OuYangPaste** [xmake-io/xmake#2412] [General]
### 关于 local ccache 有两个小疑惑:
>1. 我直接在我项目的基础上 xmake f ... --ccache=n -c, 发现 ccache 没能关掉，追踪到 $(programdir)/modules/private/cache/build_cache.lua 里发现 is_enabled 里先判断的 project.policy("build.ccache")，再判断的 config.get("ccache")，一般来说不是应该通过命令行设置的优先级会更高才对么
>2. 内置 ccache 的默认路径为 build/.build_cache, 为什么不选择像外置的 ccache 那样默认放在如 ~/.xmake 的路径里进行全局共享呢，这样的话在类似于多个地方使用 git submodules 引用同一 repo 的情况下也能很好地进行编译构建加速

**Comment by waruqi** [comment]
> 我直接在我项目的基础上 xmake f ... --ccache=n -c, 发现 ccache 没能关掉，追踪到 $(programdir)/modules/private/cache/build_cache.lua 里发现 is_enabled 里先判断的 project.policy("build.ccache")，再判断的 config.get("ccache")，一般来说不是应该通过命令行设置的优先级会更高才对么

如果你打算动态切 ccache 状态，那就压根没必要设置 `project.policy("build.ccache")` ，也就不会存在这个问题。。既然你都打算通过 project.policy("build.ccache") 设置来强行开启 ccache了，干嘛还要再命令行去关掉它。。

> 内置 ccache 的默认路径为 build/.build_cache, 为什么不选择像外置的 ccache 那样默认放在如 ~/.xmake 的路径里进行全局共享呢，这样的话在类似于多个地方使用 git submodules 引用同一 repo 的情况下也能很好地进行编译构建加速


大部分情况下，项目都是各自独立的，默认放本地，互不冲突，也节省磁盘空间。。比如你有 10 个项目，其中只有 10% 的代码是通用共享的，那么你剩下的 90% 缓存都放全局，浪费空间，等你这个项目不用了，清理这些缓存也不好清理，放本地，便于隔离，也便于清理删除，直接 `xmake clean --all` 就行了，不影响其他项目的缓存。

如果你这项目真的很多代码都需要共享编译，那么可以自己 xmake f --ccachedir= 配置，按需切到自己指定的全局路径。。

**Comment by waruqi** [comment]
project.policy("build.ccache") 这个策略，之前改动，里面默认配成了 true，默认值按理应该是 nil 才对，我改了下，更新到 dev 再试试

**Reply by OuYangPaste** [reply]
刚才验证过，已经可以了👌
ccache 的问题，确实一般来说共享的可能比较少，但如果考虑到 xrepo 包管理下，同一个 xmake 包不同 version 的 cache 目录是分隔的，这样子似乎不能更好地去复用 ccache，毕竟两个 version 之间的代码一般还是有很多源文件是没有被改动过的

**Reply by waruqi** [reply]
xrepo 包大部分都是cmake autoconf管理的，不会走 ccache，而且安装包 通常都是一次性的 不会频繁编译重装，也没必要上 ccache

**Reply by OuYangPaste** [reply]
嗯，了解了，谢谢解惑
